Update ptracers gen ctrl and add 1 missing diagnostics fill

jm-c opened the pull request:
## What changes does this PR introduce?
- move to pkg/ctrl the local copy of ctrl_map_ini_genarr.F from verification/global_oce_biogeo_bling/code_ad/ 
  that contains implementation ptracers generic control. Improve how it deals with general number of ptracers control.
- add missing diagnostics fill for diagnostics 'Um_ImplD' & 'Vm_ImplD' in impldiff.F (only relevant when pkg/autodiff is compiled)

## What is the current behaviour? 
- 2 versions of src file:  ctrl_map_ini_genarr.F, with reference version from pkg/ctrl missing ptracers controls.
- diagnostics for implicit vertical viscosity tendencies not filled when pkg/autodiff is compiled.

## What is the new behaviour 
- ptracers generic control in pkg/ctrl reference version.
- diagnostics 'Um_ImplD' & 'Vm_ImplD'  can also be used when pkg/autodiff ise compiled.

## Does this PR introduce a breaking change? 
no

## Other information:


## Suggested addition to `tag-index`
o pkg/ctrl:
  - move to pkg/ctrl the local copy of ctrl_map_ini_genarr.F from verification/
    global_oce_biogeo_bling/code_ad/ with ptracers generic control. Improve for 
    general number of ptracers control.
  - add missing diagnostics fill for diagnostics 'Um_ImplD' & 'Vm_ImplD' in 
    impldiff.F (otherwise not filled when pkg/autodiff is compiled).

jm-c commented:
Two comments:
1) the last commit in this PR (i.e., https://github.com/MITgcm/MITgcm/pull/447/commits/490cefc0897930d2e1243a1ec5e269f61273f27a ) could be reverted if needed. I view it at a little improvement but
it's subjective. But if we keep it, would need to add some check & stop in ctrl_check.F to avoid invalid "xx_ptrZ' ctrl file
suffix (e.g., when Z is not in 1 -> 9, currently trigger a gad error in CTRL_MAP_INI_GENARR).
2) I realised just now that there are many more local version of ctrl_map_ini_genarr.F ; again, once we converge on reference pkg/ctrl version, could update the other one.

jm-c commented:
Regarding (2) above: 
- The customized version of ctrl_map_ini_genarr.F in halfpipe_streamice & hs94.1x64x5/ (both code_ad and code_oad) are very specific, so suggest to leave them unchanged. 
-  For tutorial_global_oce_biogeo/ it seems it should not be difficult to use the new reference version, and may be also switch off ECCO_CTRL_DEPRECATED there. @mjlosch may be if you feel like taking a look at this one (since you already removed this option from several experiments). If we keep tutorial_dic_adjoffline/ unchanged (that use old CTRL settings, similar set of pkgs + more cusomized cost & ctrl code) this would leave us with both the gen-ctrl and non gen-ctrl tested in our 2 AD dic experiments. 
- And I managed to remove the local version in tutorial_tracer_adjsens/code_ad (was easy) and use the new reference version there. Now looking at the OpenAD test there.

And unrelated issue: should we add a (CPP) option like "ALLOW_PTR0_CONTROL" (like we have ALLOW_TR10_CONTROL
in the non gen-ctrl case for pTr 1) ?

mjlosch commented on pkg/ctrl/ctrl_map_ini_genarr.F:
```diff
@@ -35,127 +46,166 @@ SUBROUTINE CTRL_MAP_INI_GENARR( myThid )
+# endif
+
+# ifdef ALLOW_PTRACERS
+CADJ STORE pTracer=comlev1,key=ikey_dynamics,kind=isbyte
```
I don't think that this directive is necessary.  Have you tried without it?

jm-c replied:
I have not tried, I just took it from global_oce_biogeo_bling/code_ad version (from @mmazloff ). Will check.

mjlosch commented:
@jm-c 
> For tutorial_global_oce_biogeo/ it seems it should not be difficult to use the new reference version, and may be also switch off ECCO_CTRL_DEPRECATED there. @mjlosch may be if you feel like taking a look at this one (since you already removed this option from several experiments). If we keep tutorial_dic_adjoffline/ unchanged (that use old CTRL settings, similar set of pkgs + more cusomized cost & ctrl code) this would leave us with both the gen-ctrl and non gen-ctrl tested in our 2 AD dic experiments.

Do you think that this should be done in this PR? If yes, I can give it a try.

mjlosch commented:
> And unrelated issue: should we add a (CPP) option like "ALLOW_PTR0_CONTROL" (like we have ALLOW_TR10_CONTROL
in the non gen-ctrl case for pTr 1) ?

I haven't thought about this much and maybe I am overlooking something here, but I would refrain from adding anything to the deprecated code. What would be the advantage?

jm-c commented:
@mjlosch regarding a CPP option like "ALLOW_PTR0_CONTROL": it would not be for the deprecated code but rather for the new,
gen-ctrl part. The only reason I see would be to allow to have TAF view ptracers as non active. With large number of tracers (but none active) this might simplify (+ reduce size) of the code.

jm-c commented:
@mjlosch regarding other customized version of ctrl_map_ini_genarr.F : I did not manage to get the OpenAD tutorial_tracer_adjsens to work with new reference version, so give up on this one. But I was going to remove the local version
in code_ad since it works well with new ref. version. 
You could try to see how it works for tutorial_global_oce_biogeo ; and before pushing to this PR, we could decide:
a) if only minimal changes are needed then add to this one ; b) if more substantial changes then make a new PR.

jm-c commented:
I removed the TAF store directive (turns out that with TAF directive " loop = parallel" it's enough to avoid recomputation) and switched experiment tutorial_tracer_adjsens to use the new reference version of "ctrl_map_ini_genarr.F" (for TAF adjoint tests, not with OpenAD).

mjlosch commented:
@jm-c 
> You could try to see how it works for tutorial_global_oce_biogeo ; and before pushing to this PR, we could decide:
a) if only minimal changes are needed then add to this one ; b) if more substantial changes then make a new PR.

To make is run and pass the test I have to change 3 files (CTRL_OPTIONS.h, data.ctrl, data.grdchk) and add three (CTRL_SIZE.h, genweights.py, ones64.bin) and get an extra (benign) recomputation warning (xx_gen2dtim_dummy), that won't even show up in the number of warnings reported. I guess that qualifies as "only minimal changes".

But the tests do not catch that the sensitivities are actually different.
1. control variables like "ALLOW_HFLUX0_CONTROL", etc. are not supported by the generic controls and this may lead to different sensitivities in adxx_hflux (old) compared with adxx_qnet (generical control), similar to global_ocean.cs32x15 and especially since this experiment does not use pkg/cal we might not expect these sensitivities to be the same 
2. 3D fields like adxx_theta are (very) different at individual points below level 2, probably near the bottom.
3. 3D adxx_ptr is very different from adxx_tr1, because it's just 0.

That means, more work and probably a different PR.

jm-c commented:
@mjlosch two comments:
1) Should we worry that adxx_ptr is all zero ?
2) for these stupid weight files (i.e., full of "one"), it would be nice to have them called the same way across different experiments,
and suffix ".bin" is also nice (as most other binary input files).
I global_oce_biogeo_bling it's "wt_ones.bin" ; the new ones you added in PR #406 are named: ones.prec64 & ones.prec32 ; and here you are suggesting "ones64.bin" ; with "git mv" we could easyly rename the previous one once we agree on a file name.
And the other thing is that sicne these files are a bit stupid (and might disapear once we change the default weight), we could always use the same one (from the largest experiment, think is currently isomip) and make a symbolic link with prepare_run (as I did in this PR for tutorial_tracer_adjsens), but it does not prevent us to keep/add genweights.py with the right size.

mjlosch commented:
1. yes, I need to check xx_ptr that before we can change this experiment, that's why I think it should wait.
2. I am sorry, I didn't pay attention to the weight file naming convention, that was not intentional. In the current experiment I just copied from global_ocean.cs32x15 and stuck ones.prec64, I am sorry for citing the wrong name in my previous comment (it was just wrong in the comment). I think it is useful to have the file precision in the name here (e.g. ones.prec64.bin or wt_ones.prec64.bin, or wt_ones64.bin, whatever you prefer), because it really matters, but other than that I am fine with your suggestion ("*.bin"). Having on generic file with (too) many ones may be a little confusing, but I agree that this would save some space.  I am not sure how simple it will be to get rid of these files, because specifying the weights file actually turns on the generic control variable. It would be good to understand why @gaelforget decided to do it this way.
3. I haven't though about the ALLOW_PTR0_CONTROL flag. If you think it is useful, why not just add it (or show me in a non-PR branch how you would do that) and then it would be easier for me to comment (if it's not too much trouble to do so).

jm-c commented:
@mjlosch 1) regarding ALLOW_PTR0_CONTROL, I don't think what I wrote earlier make much sense, so suggest we keep it the way it is without a new CPP option. If later on there is a need for it we could had it then.
2) regarding weight file name, my favorite would be "ones_64b.bin" (with 64 bits "one") and "ones_32b.bin" (32 bits one);
ones_prec64.bin & ones_prec32.bin are fine too (and I prefer with "_" instead of "." after "ones", because it really the type of "ones" that are in the file); i don't see the need to have "wt_" (yes, it's used for weight but it's independent of the file content).

mjlosch commented:
OK, I agree with this naming convention. Do you want to do this in the PR (and do you want me to do it), or should be do this later?

gaelforget commented:
> 2\. I am not sure how simple it will be to get rid of these files, because specifying the weights file actually turns on the generic control variable. It would be good to understand why @gaelforget decided to do it this way.

Not sure but one thing I recall is that at the time I was using the generic controls as a safe way to diagnose sensitivity (adxx) vs the ADJ output (hard-coded ~ a priori assumed broken until proven otherwise). Maybe this was made easier with the quoted approach?

gaelforget commented:
> 2\. for these stupid weight files (i.e., full of "one"),

Might prefer silly over stupid as a qualifier, lol

ps. I have never created files that compress so effectively btw!

mjlosch commented:
@gaelforget thanks for your insights. Do you agree with changing this behavior, i.e., if no (silly) weights are specified, the default weights are then set to one, and the variable is turned on in a different way (by specifying the name)? Or do you advise to keep it the way it is?

jm-c commented:
I can do this in the same PR. Thanks.

On Wed, Mar 24, 2021 at 10:11:01AM -0700, Martin Losch wrote:
> OK, I agree with this naming convention. Do you want to do this in the PR (and do you want me to do it), or should be do this later?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/447#issuecomment-806004319

jm-c commented:
@gaelforget Thanks for the insight.
I am in favor of leaving this default  weight code unchanged in this PR and put this modification in an other/later PR.
Going to rename these ultra simple files.

gaelforget commented:
> @gaelforget thanks for your insights. Do you agree with changing this behavior, i.e., if no (silly) weights are specified, the default weights are then set to one, and the variable is turned on in a different way (by specifying the name)? Or do you advise to keep it the way it is?

I would need to think more about this but a priori I like your idea of using the name in place of the file name. Postponing like @jm-c suggested sounds good

jm-c commented:
I think I am done with changes here. If/when @mjlosch you have time to take a second look at latest version, we could moved towards merging after that.

mjlosch commented on pkg/ctrl/ctrl_check.F:
```diff
@@ -82,82 +96,82 @@ SUBROUTINE CTRL_CHECK( myThid )
+          errCount = errCount + 1
+        ENDIF
+      ENDDO
+#endif /* ALLOW_GENARR3D_CONTROL */
```
This is really minor: here you added a comment to the terminating endif, but earlier and later you didn't. This does not need to change, unless you want to.

jm-c replied:
I tend to add a comment after "#endif" when the #ifdef / #endif block is not too short (and in this case, this one got longer after I added these ptracers stuff). But it's still a bit arbitrary. Anyway, I am going to add few more comments after #endif in this file (specially when there are several nested #ifdef).

mjlosch commented:
LGTM

jm-c commented:
Will merge this later today (if no one complains)
